ckanext/canada/harvesters.py

```diff
@@ -0,0 +1,616 @@
+        # relatively accurate. Going back a little earlier, just in case.
+        get_changes_since = (last_time - datetime.timedelta(hours=1)).isoformat()
+        log.info('Searching for activity since: %s UTC', get_changes_since)
+
+        data = self.registry.action.changed_packages_activity_timestamp_since(since_time=last_time.isoformat())
+        if not data:
+            # There is no activity since last successful run
+            return []
+
+        registry_packages = []
+
+        for result in data:
+            package_id = result['package_id']
+            try:
+                source_package = self.registry.action.package_show(id=package_id)
+            except plugins.toolkit.ObjectNotFound:
+                log.info("Package %s not found in Registry database.", package_id)
+                self._save_gather_error("Package (%s) not found in Registry database.", package_id)
+                pass
+            else:
+                source_package = self._get_datastore_and_views(source_package, ckan_instance=self.registry, stage='GATHER')
+                registry_packages.append(source_package)
+
+        harvest_object_ids = []
+
+        for registry_package in registry_packages:
+            source_package = registry_package.copy()
+
+            if source_package and (source_package.get('state') == 'deleted' or source_package.get('type') not in SYNC_DATASET_TYPES):
+                # Do not sync deleted packages or packages that are NOT type: 'info', 'dataset', 'prop'
+                source_package = None
+
+            _trim_package_for_sync(source_package)
+
+            if source_package:
+                now = datetime.now()
+                if source_package.get('ready_to_publish') == 'false':
+                    source_package = None
+                    log.info("Package %s marked not ready to publish.", source_package['id'])
+                elif not source_package.get('portal_release_date'):
+                    source_package = None
+                    log.info("Package %s release date not set.", source_package['id'])
+                elif isodate(source_package.get('portal_release_date'), None) > now:
+                    source_package = None
+                    log.info("Package %s release date in future.", source_package['id'])
+                else:
+                    # Portal packages are public, do this to match
+                    source_package['private'] = False
+
+            log.info('Creating HarvestObject for package %s', registry_package['id'])
+            obj = HarvestObject(guid=registry_package['id'],
+                                harvest_source_id=harvest_job.source.id,
+                                harvest_job_id=harvest_job.id,
+                                job=harvest_job,
+                                package_id=registry_package['id'],
+                                content=json.dumps(source_package) if source_package else None)
```

Comment:
It seems weird to be getting storing the data in the gather stage and not the fetch stage. Also, copying this data into a new table just because the harvester uses separate stages seems like extra work for no benefit.

Reply:
Yeah from the docs and code of the ckanext-harvest plugin I did not really like how it stores so much in the DB tables. We can probably do everything in the gather stage in terms of getting the Registry and Portal packages and comparing them.

I imagine that the gather_errors will still show us something at least in the UI and database.

The framework of the harvest plugin does kind of require us to actually use the import_stage though. So we would need to either store the data in the DB tables to use in the import_stage, or just go get everything again. Which would be annoying that it does everything again, but I don't think it would hurt that much. I would rather have it go get the Registry package, views, and data dictionary again than store giant objects in the database.

I think another issue is how the fetch_stage works. From what I can tell, the harvest plugin will do stuff and show us things for the return of the fetch. I guess we could save a small object in the DB from gather_stage. Not the whole objects, but some small thing like `{"state": "unchanged"} | {"state": "success"} | {"state": "failed"}`. Will check out what the column is for the object content, might not need to JSON it.

Reply:
Chaos, I have moved the HarvestObjectExtras to just the content here now. So we have the status of the fetch stage, the import action, and the resource hashes.